VerifyError: Instruction type does not match stack map - Regression in String-switch dispatch for methods exceeding 32 KB

After recently upgrading from eclipse 2024-6 to 2025-12 some programs using auto generated code (classic deteministic code generation, not AI) started failing with:
```
Error: Unable to initialize main class BS
Caused by: java.lang.VerifyError: Instruction type does not match stack map
Exception Details:
  Location:
    BS.run()I @119: istore_2
  Reason:
    Current frame's stack size doesn't match stackmap.
  Current Frame:
    bci: @119
    flags: { }
    locals: { 'BS', integer, integer, 'java/lang/String' }
    stack: { integer }
  Stackmap Frame:
    bci: @119
    flags: { }
    locals: { 'BS', integer, integer, 'java/lang/String' }
    stack: { }
  Bytecode:
    0000000: 033c 033d 2ab4 000d 594e b600 16ab 0000
[...]
    0008650: c800 0000 091b 1c60 acc8 ffff 79ab     
  Stackmap Table:
    append_frame(@4,Integer,Integer)
    append_frame(@40,Object[#23])
    same_frame(@54)
    same_frame(@59)
    same_frame(@73)
    same_frame(@78)
    same_frame(@119)
    same_frame(@138)
    same_frame_extended(@34389)
    chop_frame(@34393,1)
```
After some investigation it turns out that it only seems to occur when using a switch case in a method exceeds a certain size. The example [BS.java](https://github.com/user-attachments/files/26966666/BS.java) works in Eclipse Version: 2024-12 (4.34.0) where it prints 413914 to the console. However starting with Eclipse Version: 2025-03 (4.35.0) produces the error.

This patch fixes the issue and includes regression tests. [ecj-goto-w-fix.patch](https://github.com/user-attachments/files/26967271/ecj-goto-w-fix.patch) AI Disclaimer: This patch was created using copilot. I am not familiar enough with this codebase to judge if this is the correct way to fix the issue. Further copilot identified this commit as cause of the regression https://github.com/eclipse-jdt/eclipse.jdt.core/commit/57fb40738b9f5b79a680509c95b9f661e3850ade

--- a/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/SwitchStatement.java
+++ b/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/SwitchStatement.java
@@ -925,7 +925,7 @@ public class SwitchStatement extends Expression {
 				this.swich.breakLabel.initialize(codeStream);
 				this.caseLabels = new BranchLabel[this.swich.nConstants];
 				gatherLabels(codeStream, BranchLabel::new);
-				this.swich.defaultLabel = new CaseLabel(codeStream, true /* allow narrow branch to */);
+				this.swich.defaultLabel = new CaseLabel(codeStream, true /* reachable also via goto[_w] */);
 				if (this.swich.defaultCase != null)
 					this.swich.defaultCase.targetLabel = this.swich.defaultLabel; // Replace the vanilla branch label with a case label that doubles as a branch label.
 			}
@@ -980,7 +980,7 @@ public class SwitchStatement extends Expression {
 					if (i == 0 || hashCode != lastHashCode) {
 						lastHashCode = hashCode;
 						if (i != 0)
-							codeStream.goto_(this.swich.defaultLabel);
+							codeStream.goto_(this.swich.defaultLabel.branchLabel);
 						this.caseLabels[j++].place();
 					}
 					codeStream.load(this.swich.selector);
@@ -988,7 +988,7 @@ public class SwitchStatement extends Expression {
 					codeStream.invokeStringEquals();
 					codeStream.ifne(this.stringCaseConstants[i].label);
 				}
-				codeStream.goto_(this.swich.defaultLabel);
+				codeStream.goto_(this.swich.defaultLabel.branchLabel);
 			}
 		}
 

--- a/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/codegen/CaseLabel.java
+++ b/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/codegen/CaseLabel.java
@@ -16,15 +16,15 @@ package org.eclipse.jdt.internal.compiler.codegen;
 public class CaseLabel extends BranchLabel {
 
 	public int instructionPosition = POS_NOT_SET;
-	private BranchLabel branchLabel; // doppelganger
+	public BranchLabel branchLabel; // doppelganger label useful in situations where the location can be reached via a goto or goto_w. The CaseLabel itself is reachable only by switch
 
 public CaseLabel(CodeStream codeStream) {
 	super(codeStream);
 }
 
-public CaseLabel(CodeStream codeStream, boolean allowNarrowBranch) {
+public CaseLabel(CodeStream codeStream, boolean reachableViaGoto) {
 	super(codeStream);
-	if (allowNarrowBranch)
+	if (reachableViaGoto)
 		this.branchLabel = new BranchLabel(codeStream);
 }
 

--- a/org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/XLargeTest.java
+++ b/org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/XLargeTest.java
@@ -2041,6 +2041,49 @@ public void testIssue1359() {
 			"The operand stack is exceeding the 65535 bytes limit\n" +
 			"----------\n");
 }
+// https://github.com/eclipse-jdt/eclipse.jdt.core/issues/5020
+// VerifyError: Instruction type does not match stack map - Regression in String-switch dispatch for methods exceeding 32 KB
+public void testIssue5020() {
+	StringBuilder sourceCode = new StringBuilder(
+			"""
+			public class X {
+			    public int run(String label) {
+			        int x = 0; int y = 0;
+			        loop: while (true) {
+			            switch (label) {
+			                default:
+			                case "A":
+
+			""");
+
+	for (int i = 0; i < 49; i++) {
+		for (int j = 0; j < 100; j++)
+			sourceCode.append("\t\t		x += " + j + "; y = x ^ y;\n");
+	}
+	sourceCode.append(
+			"""
+			    				label = "B";
+			                    continue loop;
+			                case "B":
+			                    return x + y;
+			            }
+			        }
+			    }
+			    public static void main(String[] args) {
+			        System.out.println(new X().run("A"));
+			    }
+			}
+			""");
+
+
+	this.runConformTest(
+			new String[] {
+					"X.java",
+					sourceCode.toString()
+			},
+			"413914");
+
+}
 public static Class testClass() {
 	return XLargeTest.class;
 }

